src/uct/ugni/smsg/ugni_smsg_iface.c

```diff
@@ -164,12 +164,17 @@ UCS_CLASS_DEFINE_DELETE_FUNC(uct_ugni_smsg_iface_t, uct_iface_t);
 static void uct_ugni_smsg_progress(void *arg)
 {
     uct_ugni_smsg_iface_t *iface = (uct_ugni_smsg_iface_t *)arg;
+    uint32_t attempts;
 
     progress_local_cq(iface);
     progress_remote_cq(iface);
 
     /* have a go a processing the pending queue */
-    ucs_arbiter_dispatch(&iface->super.arbiter, 1, uct_ugni_ep_process_pending, NULL);
+    iface->super.send_in_pending = 1;
+    attempts = ucs_min(iface->config.smsg_max_credit, iface->super.send_q_len);
+    ucs_arbiter_dispatch(&iface->super.arbiter, attempts,
+                         uct_ugni_ep_process_pending, NULL);
+    iface->super.send_in_pending = 0;
```

Comment:
It looks like a hack + extra instructions in progress and brunch in post send. We have to avoid this all together. I suggest to step back and understand why we have the problem all together and how it was fixed in other transports.